File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/balancedTree.ts
import {arrayToBTLevelOrder, BTNode} from "./BinaryTreeOperations";

function checkBalancedTree(root: BTNode) {
	return heightCheck(root) !== -1;
}

function heightCheck(root: BTNode) {
	if (root === null) return 0;
	let l = heightCheck(root.left);
	if (l === -1) return -1;
	let r = heightCheck(root.right);
	if (r === -1) return -1;

	if (Math.abs(l - r) > 1) return -1;
	return Math.max(l, r) + 1;
}

let root = arrayToBTLevelOrder([1,2,3,4,5,6]);
// let root = new BTNode(1);
// root.left = new BTNode(2);
// root.right = new BTNode(3);
// root.left.left = new BTNode(4);
// root.left.left.left = new BTNode(5);

console.log(checkBalancedTree(root));

File: StriverA2Z/Java/BinaryTrees/BalancedTree.java
public class BalancedTree {
    public static void main(String[] args) {
        Node root = BinaryTreeOperations.arrayToBTLevelOrder(new Integer[]{1, 2, 3, 5, 6});
//        Node root = new Node(1);
//        root.left = new Node(2);
//        root.right = new Node(3);
//        root.left.left = new Node(4);
//        root.left.left.left = new Node(5);

        System.out.println(checkBalancedTree(root));
    }

    public static boolean checkBalancedTree(Node root) {
        return heightCheck(root) != -1;
    }

    public static int heightCheck(Node root) {
        if (root == null) return 0;
        int lh = heightCheck(root.left);
        if (lh == -1) return -1;
        int rh = heightCheck(root.right);
        if (rh == -1) return -1;
        if (Math.abs(lh - rh) > 1) return -1;
        return Math.max(lh, rh) + 1;
    }
}
